packages/@aws-cdk/aws-cloudwatch/lib/metric.ts

```diff
@@ -224,12 +333,108 @@ export class Metric implements IMetric {
       return [];
     }
 
-    const list = Object.keys(dims).map(key => ({ name: key, value: dims[key] }));
+    const list = Object.keys(dims).sort().map(key => ({ name: key, value: dims[key] }));
 
     return list;
   }
 }
 
+/**
+ * A math expression built with metric(s) emitted by a service
+ *
+ * The math expression is a combination of an expression (x+y) and metrics to apply expression on.
+ * It also contains metadata which is used only in graphs, such as color and label.
+ * It makes sense to embed this in here, so that compound constructs can attach
+ * that metadata to metrics they expose.
+ *
+ * This class does not represent a resource, so hence is not a construct. Instead,
+ * MathExpression is an abstraction that makes it easy to specify metrics for use in both
+ * alarms and graphs.
+ */
+export class MathExpression extends BaseMetric {
+  /**
+   * The expression defining the metric.
+   */
+  public readonly expression: string;
+
+  /**
+   * The metrics used in the expression as KeyValuePair <id, metric>.
+   */
+  public readonly expressionMetrics: Record<string, IMetric>;
+
+  /**
+   * Label for this metric when added to a Graph.
+   */
+  public readonly label?: string;
+
+  /**
+   * Color for this metric when added to a Graph.
+   */
+  public readonly color?: string;
+
+  constructor(props: MathExpressionProps) {
+    super();
+    this.expression = props.expression;
+    this.expressionMetrics = props.expressionMetrics;
+    this.label = props.label;
+    this.color = props.color;
+
+    const invalidVariableNames = Object.keys(props.expressionMetrics).filter(x => !validVariableName(x));
+    if (invalidVariableNames.length > 0) {
+      throw new Error(`Invalid variable names in expression: ${invalidVariableNames}. Must start with lowercase letter and only contain alphanumerics.`);
+    }
+
+    validateNoIdConflicts(this);
+  }
+
+  /**
+   * Return a copy of Metric with properties changed.
+   *
+   * All properties except namespace and metricName can be changed.
+   *
+   * @param props The set of properties to change.
+   */
+  public with(props: MetricOptions): MathExpression {
```

Comment:
This `props` object should only contain the fields that mean something for `MathExpression`.